### PR TITLE
Python: Port reflected XSS query

### PR DESCRIPTION
**Pull request by RasmusWL**
Currently a draft PR as it builds on top of https://github.com/github/codeql/pull/4534

---

The commit history for this PR is not perfect, since there are some commits that could have been squashed with previous commits. I didn't do it, since I think it would have taken too long compared to what is gained from it.

To make things simple for the XSS query, I adopted the same approach we currently use, and treat the _creation_ of a HTTP response a sink -- and don't try to take into account whether the HTTP response is actually send back to the client.

### ConceptTests

Since I added quite a bit of tags for any HTTP response, all the other tests had to get quite a bit of annotations, that wasn't really too useful. The interesting bits of how we model HTTP responses happen in the newly added `response_test.py` file. So I came up with a neat little trick in https://github.com/github/codeql/commit/16cdfd63c6a4f7cdb0bd054e6e690094e0c7da13 to allow us to specify we only care about some subset of tags for certain files. I think it worked out pretty well :blush: 

### Modeling flask responses

In a route handler in flask, you are allowed to do return a string, and it will be wrapped into a response, like so:

```py
@app.route("/hello")
def hello(): 
    return "<h1>hello</h1>"
```

However, I was a bit sad that for the following snippet, I had to mark **both** places as a creation of a HTTP response -- in reality it is only happening with `make_response()`. More details in commit messages, but basically I gave up in the end, and just model both things as HTTP response creation (and I admit that I spend a little too much time pondering this problem).

```
resp = make_response(...)
return resp
```

### CODE REVIEW

File: python/ql/test/experimental/meta/ConceptsTest.qll [hunk 1]
```diff
@@ -142,3 +142,42 @@ class HttpServerRouteSetupTest extends InlineExpectationsTest {
     )
   }
 }
+
+class HttpServerHttpResponseTest extends InlineExpectationsTest {
+  File file;
+
+  HttpServerHttpResponseTest() { this = "HttpServerHttpResponseTest: " + file }
```
Comment (yoff) [inline python/ql/test/experimental/meta/ConceptsTest.qll:149]: Should we restrict `file` further here? Say to Python-files within the test directory?

File: python/ql/src/experimental/semmle/python/frameworks/Flask.qll [hunk 2]
```diff
@@ -131,9 +165,99 @@ private module FlaskModel {
...
+    /**
+     * A source of an instance of `flask.Response`.
+     *
+     * This can include instantiation of the class, return value from function
+     * calls, or a special parameter that will be set when functions are call by external
+     * library.
+     *
+     * Use `Response::instance()` predicate to get references to instances of `flask.Response`.
+     */
+    abstract class InstanceSource extends HTTP::Server::HttpResponse::Range, DataFlow::Node { }
```
Comment (yoff) [inline python/ql/src/experimental/semmle/python/frameworks/Flask.qll:218]: I guess the concrete `InstanceSource`s are yet to be written.

File: python/ql/src/experimental/semmle/python/frameworks/Django.qll [hunk 3]
```diff
@@ -627,6 +638,905 @@ private module Django {
...
+            t.start() and
+            result = response_attr("HttpResponse")
+            or
+            // TODO: remove/expand this part of the template as needed
+            // Handle `http.HttpResponse` alias
+            t.start() and
+            result = http_attr("HttpResponse")
+            or
+            // subclass
+            result.asExpr().(ClassExpr).getABase() = classRef(t.continue()).asExpr()
```
Comment (tausbn) [inline python/ql/src/experimental/semmle/python/frameworks/Django.qll:714]: I feel like we should just include this in all of the models of  classes that extend `HttpResponse`. Would this not do what we want?

File: python/ql/src/experimental/semmle/python/frameworks/Django.qll [hunk 4]
```diff
@@ -627,6 +638,905 @@ private module Django {
...
+        module HttpResponseRedirect {
+          /** Gets a reference to the `django.http.response.HttpResponseRedirect` class. */
+          private DataFlow::Node classRef(DataFlow::TypeTracker t) {
+            t.start() and
+            result = response_attr("HttpResponseRedirect")
+            or
+            // TODO: remove/expand this part of the template as needed
+            // Handle `http.HttpResponseRedirect` alias
+            t.start() and
+            result = http_attr("HttpResponseRedirect")
```
Comment (tausbn) [inline python/ql/src/experimental/semmle/python/frameworks/Django.qll:782]: ... so, here, for instance, to capture subclasses of `HttpResponseRedirect`.

File: python/ql/src/experimental/semmle/python/frameworks/Flask.qll [hunk 5]
```diff
@@ -131,9 +165,99 @@ private module FlaskModel {
 
       /** Gets a reference to the `add_url_rule` method on an instance of `flask.Flask`. */
       DataFlow::Node add_url_rule() { result = instance_attr("add_url_rule") }
+
+      /** Gets a reference to the `make_response` method on an instance of `flask.Flask`. */
+      // HACK: We can't call this predicate `make_response` since shadowing is
+      // completely disallowed in QL. I added an underscore to move thing forwards for
```
Comment (tausbn) [inline python/ql/src/experimental/semmle/python/frameworks/Flask.qll:171]:
```suggestion
      // completely disallowed in QL. I added an underscore to move things forward for
```

**Review by tausbn — CHANGES_REQUESTED** [review]
I think overall this looks good and close to mergeable. I think we can reuse the subclass type tracking pattern a few more places, to extend the modelling.

File: python/ql/src/experimental/semmle/python/frameworks/Django.qll (hunk 3 above)
Comment (yoff) [reply python/ql/src/experimental/semmle/python/frameworks/Django.qll:714]: Yes, I think you are right, and it is not a lot of code to duplicate...

**Review by tausbn — DISMISSED** [review]
Awesome!

**Review by tausbn — APPROVED** [review]

**Pull request merged**